apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryDataReader.cs

```diff
@@ -14,197 +14,209 @@
 
 using System;
 using System.Collections;
-using System.Data;
+using System.Collections.Generic;
 using System.Data.Common;
+using System.Globalization;
+using System.Threading;
+using System.Threading.Tasks;
+using Google.Api.Gax;
 
 namespace Google.Cloud.BigQuery.V2
 {
     /// <summary>
+    /// Reads a forward-only stream of rows from a data source.
     /// </summary>
     public sealed class BigQueryDataReader : DbDataReader
     {
-#if NET45
+        private readonly BigQueryConnection _connectionToClose;
+        private readonly IAsyncEnumerator<BigQueryRow> _results;
+        private bool _isEnumerating;
+
+        internal BigQueryDataReader(BigQueryResults bigQueryResults, BigQueryConnection connectionToClose = null)
+        {
+            _connectionToClose = connectionToClose;
+            GaxPreconditions.CheckNotNull(bigQueryResults, nameof(bigQueryResults));
+            BigQueryResults = bigQueryResults;
+            _results = BigQueryResults.GetRowsAsync().GetEnumerator();
+        }
+
 
+#if NET45 
         /// <inheritdoc />
         public override void Close()
         {
-            throw new NotImplementedException();
+            _results?.Dispose();
+            _connectionToClose?.Close();
         }
 
         /// <inheritdoc />
-        public override DataTable GetSchemaTable()
+        public override System.Data.DataTable GetSchemaTable()
         {
-            throw new NotImplementedException();
+            // We cannot throw here as some scenarios (datatable) will call this method
+            return null;
         }
 #endif
 
         /// <inheritdoc />
-        public override bool NextResult()
-        {
-            throw new NotImplementedException();
-        }
+        public override bool NextResult() => false;
 
         /// <inheritdoc />
         public override bool Read()
-        {
-            throw new NotImplementedException();
-        }
+            => ReadAsync(CancellationToken.None).ResultWithUnwrappedExceptions();
 
         /// <inheritdoc />
-        public override int Depth { get; }
+        public override async Task<bool> ReadAsync(CancellationToken cancellationToken)
+            => await MoveNextAsync(cancellationToken).ConfigureAwait(false);
+
         /// <inheritdoc />
-        public override bool IsClosed { get; }
+        public override int Depth => 0;
+
+        //HasRows is true before any MoveNext and thereafter as long as Current is not null.
         /// <inheritdoc />
-        public override int RecordsAffected { get; }
+        public override bool HasRows => !_isEnumerating || _results.Current != null;
```

Comment:
I've been thinking about this more, and it feels like we should return the right answer, even if that means making an RPC. That's ugly in terms of maybe needing to buffer the results of a MoveNext call, but I don't think returning the wrong answer is a good idea.